### ISSUE TITLE
As a task creator I need to get notification when a comment is added to my task.

### ISSUE THREAD

**Issue by LisaLNelson** [openopps/openopps-platform#1236]
It appears that the notification to the task creator when a comment has been posted to their task has been turned off. For example:
![image](https://cloud.githubusercontent.com/assets/9039852/12625401/a14ad766-c501-11e5-8170-96b534b2f941.png)
- [ ] make a failing test
- [ ] fix the bug

The acceptance criteria for this story is to just fix the comment notification. Ignore the comments below about threading. We can add a separate story and prioritize for later

**Comment by ultrasaurus** [comment]
@rogeruiz this feels like it is a regression... do you think you could take a quick look and see if it might have been introduced when you made notification changes for 'save as draft' improvements?

**Comment by Connery-LCSolutions** [comment]
@LisaLNelson I'm having a hard time reproducing this error on the test server. Attached is the notification that I receive when I comment on a task that I created. Are you part of the Open Opps Google Group? I'm about to post something longer, but in short, I'm wondering if notifications for you are working, but are getting blocked/deleted as spam. 

http://stackoverflow.com/questions/29027370/mandrill-messages-moved-to-spam-because-of-caution-message-contains-redirect

![replyemail](https://cloud.githubusercontent.com/assets/15251877/13907141/4aec9fb2-eebe-11e5-96a5-cf8cc1bf02dc.PNG)

**Comment by LisaLNelson** [comment]
When someone posts a comment on a task the person posting the task and me as the admin are getting the message below, but the person creating the task doesn't. Bernetta Reese is the creator of the task that this comment was posted on.  You can see she is not included.

![screen shot 2016-03-20 at 5 22 59 pm](https://cloud.githubusercontent.com/assets/9039852/13907265/c035ee3e-eec0-11e5-9be0-ce41243df380.png)

**Comment by LisaLNelson** [comment]
If you send me the link to the staging server I will try to recreate it there.

**Comment by Connery-LCSolutions** [comment]
Is there is additional staging server? I was using the test server here: https://openopps-test.18f.gov/

**Comment by Connery-LCSolutions** [comment]
@LisaLNelson Ok, I understand. If you have a second, would you please leave a comment on the new task I created on the test server? Perhaps I was getting the notification email because I was the poster, as well as the task creator.

**Comment by ultrasaurus** [comment]
@Connery-LCSolutions you won't get notifications if you aren't registered with your real email -- sorry I thought the test server was configured to allow anyone.  I'll look at fixing that now!

**Comment by Connery-LCSolutions** [comment]
@ultrasaurus Right, I understand. Just posted my working theory in Google Group. Am proceeding to try and test that now.

**Comment by ultrasaurus** [comment]
Adding a note here:  test server updated to allow anyone to create an account

**Comment by Connery-LCSolutions** [comment]
@ultrasaurus Can't believe you both are working on a Sunday. 18F is like the Special Forces of coding!

**Comment by ultrasaurus** [comment]
Aside from what I expect is an unrelated issue with profile images being broken on the site, Sarah Allen was able to comment on an opportunity posted by Sarah Saurus:  https://openopps-test.18f.gov/tasks/67  (and Sarah Saurus still owns it)

I posted the opportunity (then approved it with a different adminstrator account) and received notifications:
![image](https://cloud.githubusercontent.com/assets/41906/13908427/83949940-eebf-11e5-8797-7cdc68801fd0.png)

Then Sarah Allen commented on the opportunity and received this email:
![image](https://cloud.githubusercontent.com/assets/41906/13908459/feace740-eebf-11e5-8284-adb6bea6cc4c.png)


which I think was supposed to be received by Sarah Saurus.

**Comment by Connery-LCSolutions** [comment]
@ultrasaurus I haven't been getting any notifications for the last two opportunities to either account. Is this a known issue, or is the mail server down now?

**Comment by LisaLNelson** [comment]
I posted  a couple comments on the staging server and did not get notifications.

**Comment by ultrasaurus** [comment]
@Connery-LCSolutions I did not see a change of ownership.  I'm getting email notifications from https://openopps-test.18f.gov in general, but not when someone comments on an opportunity that I posted.

**Comment by Connery-LCSolutions** [comment]
@ultrasaurus Understood and Agreed. I miscommunicated, in that, I don't think ownership is actually changing. I would like to know if, be design, the Commenter is suppose to be receiving that email however.  Unfortunately, the test server seems to have stopped sending all comments last night at around 9, and it does not seem to be working again (I just tested with a new comment). If you decide to reboot the server, may I recommend starting with the mail system first, and letting us test notifications before rebooting Open-Opps. While, I don't think the mail delivery system is the cause of the Commenter notification issue, it would provide insight into why all notifications stopped last night. Nodemailer should be able to handle far more than 3 concurrent users, so I would be surprised if it was an architectural design error. 

I think this issue goes deeper than Commenter notifications (or there are more notification issues than just associated with the Commenter).

**Comment by ultrasaurus** [comment]
@Connery-LCSolutions I will look at the configuration of the test server.  Did you ever get it running locally?  I think it would be much easier to debug if you looked at the local log file.  I believe it reports outgoing messages in the log (even if you don't have a mail server configured) -- or you could add whatever instrumentation would be helpful.

**Comment by Connery-LCSolutions** [comment]
@ultrasaurus I've been working out of the docker container since I only had 5 days. I assumed (my mistake) that the outgoing messages would have showed up in the container logs. Is there a way to access additional logs within the docker container. If not, I can finish my local dev environment today.

**Comment by ultrasaurus** [comment]
@Connery-LCSolutions you should see the app log in the console

**Comment by Connery-LCSolutions** [comment]
@ultrasaurus Yes, Ma'am. The email log for Docker is attached. While it doesn't tell us a lot, it does show that notifications are being sent out. I performed six tasks and it sent attempted to send out six notifications. Unfortunately, we can't tell if the messages for 5 and 6 would have gone to User A or B. With additional time, I could add additional logger statements for the recipient. 

![ninjas](https://cloud.githubusercontent.com/assets/15251877/13923483/309f0250-ef57-11e5-8525-9627a6ef9c49.PNG)

1. Created New User A
2. Posted new opportunity (User A is now Task Owner)
3. User A commented on own opportunity.
4. Created New User B 
5. Posted Comment 1 from New User B
6. Posted Comment 2 from New User B 
![mailerror](https://cloud.githubusercontent.com/assets/15251877/13923260/397fb9ba-ef56-11e5-9cfc-b2d257b36388.PNG)

**Comment by Connery-LCSolutions** [comment]
Note that User A (Daniel Connery) created the task and should therefore be the Task Owner, but isOwner is set to False. User B (Danielle Connery) is the Commenter, but isOwner is set to True. 

![isowner](https://cloud.githubusercontent.com/assets/15251877/13924172/b6b73aa4-ef59-11e5-849c-a8aac1c5f980.PNG)

**Comment by ultrasaurus** [comment]
I believe the root of the bug is this line:
https://github.com/18F/openopps-platform/blob/dev/api/notifications/comment.create.owner/notification.js#L5

where the owner is set as the person who made the comment here:
https://github.com/18F/openopps-platform/blob/dev/api/notifications/comment.create.owner/notification.js#L31

this was pointed out recently on the Google Group by @Connery-LCSolutions
```
I'm curious to see if it's related to the following code.

 // TODO: Does this need two references to the `commenter` object?
      // Explore how `data` is being used within the Notification template and refactor, maybe.
      data.owner = data.commenter = commenter;

I'm not sure I understand the author's intent here, but if it's assigning the "Commenter" as the new data owner, that would certainly explain the issue. 
```

We still need to create a failing test for this case -- none of the Notifications have tests and it's hard to catch these kinds of issues interactively.

**Comment by Connery-LCSolutions** [comment]
@ultrasaurus Creating a failing test to verify the original owner is simple enough. My thought is that 2 tests would be desired. 1 for verifying the a notification was actually sent, and one to verify the correct recipient. Once this is complete, I expect you'll want a 3rd notification and corresponding test for "nested comments". Unless I missed something, there is no notification sent out when User C comments on User B's comment on User A's Task. In that use case, I would assume you would want a notification to got to User A, the task creator (the current notification would suffice), and User B, the initial commenter (a new notification would have to be written). Alternatively, 18F could just disable nested comments as a design decisions agreed upon by your Leadership. This is a great project, good luck!

**Comment by acolter** [comment]
See also https://github.com/openopps/openopps-platform/issues/1215

**Comment by mtorres253** [comment]
@LisaLNelson In our meeting today @rogeruiz asked if this is a higher priority than the full-time details work. It seems to be based on our backlog organization, but we think it might not be. Would you leave a comment here one way or the other?

**Comment by kbwagner** [comment]
@LisaLNelson - can you test this again?  I tested it in Stage/UAT last week and it worked.  Might be resolved now with the migration.

**Comment by kbwagner** [comment]
1/24/18 - Katrina tested this in Stage and it appears to be fixed.  Perhaps was fixed with the migration.  Test in production and then close if successful.

**Comment by LisaLNelson** [comment]
1/24/18 Lisa Reviewed
